ngramoffset.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2021 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zoekt
+
+import (
+	"log"
+	"reflect"
+	"sort"
+)
+
+// simple interface type check for implementations
+var (
+	_ ngramOffset = mappedNgramOffset{}
+	_ ngramOffset = &arrayNgramOffset{}
+)
+
+type ngramOffset interface {
+	Get(ngram) simpleSection
+	DumpMap() map[ngram]simpleSection
+	Len() int
+}
+
+func makeNgramOffsetEmpty() ngramOffset {
+	return mappedNgramOffset(map[ngram]simpleSection{})
+}
+
+func makeNgramOffset(ngrams map[ngram]simpleSection) ngramOffset {
+	if len(ngrams) > 0 {
+		a := makeArrayNgramOffset(ngrams)
+		m := a.DumpMap()
+		if !reflect.DeepEqual(m, ngrams) {
```

Comment:
Do we need to have this assertion on the critical path or could we just rely on tests for the same thing?

Reply:
I removed this assertion after adding better tests.